Added xdg_surface

TheNeikos opened the pull request.

TheNeikos committed:
Added xdg_surface implementation

TheNeikos committed:
Added sample xdg_shell implementation

TheNeikos committed:
Added xdg_shell and xdg_surface to comp module

neithernut commented:
Mh... the setup for code generation from the XML is still missing. I'll show you how to do it on monday.

matthiasbeyer commented on src/compositor/wayland/xdg_surface.c:
```diff
@@ -0,0 +1,319 @@
+
+
+
+
```
Four blank lines are three too much!

matthiasbeyer replied:
Also check the blank lines between the functions.

TheNeikos committed:
Added a memory check for wl_region

TheNeikos committed:
Fixed missing occurrence of client resource wl_shell

TheNeikos committed:
Fixed wrong function and moved the implementation in wl_region

TheNeikos committed:
Added protocol cmake target

The protocol xml is from the weston sources, as it is licensed under MIT
we can freely include without infringing on any license.

matthiasbeyer commented on src/compositor/wayland/xdg_shell.c:
```diff
@@ -0,0 +1,247 @@
+    .dump_callback = NULL,
+    .run_callback = NULL,
+    .cmp_callback = NULL,
+};
```
Here are two things missing: The attribute table and the function table.

neithernut replied:
Will not work without rebasing. It's based on a commit in which the function table wasn't yet integrated.

matthiasbeyer replied:
So he has to rebase...

matthiasbeyer commented on src/compositor/wayland/xdg_shell.c:
```diff
@@ -0,0 +1,247 @@
+    }
+
+    struct wl_display* display = ws_wayland_acquire_display();
+
```
Please remove this blank line.

matthiasbeyer commented on src/compositor/wayland/xdg_shell.c:
```diff
@@ -0,0 +1,247 @@
+    struct wl_resource* resource
+) {
+    struct ws_wayland_xdg_shell* self = calloc(1, sizeof(*self));
+
```
Please remove this blank line.

matthiasbeyer commented on src/compositor/wayland/xdg_shell.c:
```diff
@@ -0,0 +1,247 @@
+    }
+
+    int retval = ws_wayland_obj_init(&self->wl_obj, resource);
+
```
Please remove this blank line.

matthiasbeyer commented on src/compositor/wayland/xdg_shell.c:
```diff
@@ -0,0 +1,247 @@
+    int retval = ws_wayland_obj_init(&self->wl_obj, resource);
+
+    if (retval < 0) {
+        return NULL;
```
You lose the `struct ws_wayland_xdg_shell` here.

matthiasbeyer commented on src/compositor/wayland/xdg_shell.c:
```diff
@@ -0,0 +1,247 @@
+    struct wl_resource* resource,
+    int32_t version
+) {
+
```
Missing `//!< @todo` here.

matthiasbeyer commented on src/compositor/wayland/xdg_shell.c:
```diff
@@ -0,0 +1,247 @@
+    int32_t y,
+    uint32_t flags
+) {
+
```
Missing `//!< @todo` here.

matthiasbeyer commented on src/compositor/wayland/xdg_shell.c:
```diff
@@ -0,0 +1,247 @@
+    struct wl_resource* resource,
+    uint32_t serial
+) {
+
```
Missing `//!< @todo` here.

matthiasbeyer commented on src/compositor/wayland/xdg_shell.c:
```diff
@@ -0,0 +1,247 @@
+    uint32_t version,
+    uint32_t id
+) {
+    version = 1; // We only implement version 1 right now
```
Missing `//!< @todo` here?

TheNeikos replied:
Well, there is nothing left to do, but we might support other versions later down the line

matthiasbeyer commented on src/compositor/wayland/xdg_surface.c:
```diff
@@ -0,0 +1,378 @@
+    .unset_maximized        = &unset_maximized,
+    .set_fullscreen         = &set_fullscreen,
+    .unset_fullscreen       = &unset_fullscreen,
+    .set_minimized          = &set_minimized,
```
These are all function ptrs, right? Why do you have to `&` here?

matthiasbeyer commented on src/compositor/wayland/xdg_surface.c:
```diff
@@ -0,0 +1,378 @@
+    }
+
+    struct ws_wayland_xdg_surface* self = calloc(1, sizeof(*self));
+
```
Please remove this blank line.

matthiasbeyer commented on src/compositor/wayland/xdg_surface.c:
```diff
@@ -0,0 +1,378 @@
+    }
+
+    int retval = ws_wayland_obj_init(&self->wl_obj, resource);
+
```
Please remove this blank line.

matthiasbeyer commented on src/protocol/CMakeLists.txt:
```diff
@@ -0,0 +1,19 @@
+
+
```
Please remove these blank lines.

matthiasbeyer commented:
The indentation of the XML is pretty fucked up, I guess you copy-pasted it, right? Can you please reindent it?

neithernut commented:
I would not reindent it, but keep it the way it is, as it _may_ be updated by C&Ping over a new version.

TheNeikos commented:
Yeah, I took the xml as is.

matthiasbeyer commented:
@neithernut made a good point.

TheNeikos commented:
Okay, I fixed all your comments, as you have not made them on the now modified lines they still appear as unresolved. I trust you to change that.

matthiasbeyer commented:
Feel free to do some squashing! I guess we will pair-review this evening!

neithernut commented:
ACK.

matthiasbeyer commented:
ACK.
